lms/static/scripts/frontend_apps/utils/test/use-warn-on-page-unload-test.js

```diff
@@ -0,0 +1,66 @@
+import { mount } from 'enzyme';
+import { useState } from 'preact/hooks';
+
+import { useWarnOnPageUnload } from '../use-warn-on-page-unload';
+
+describe('useWarnOnPageUnload', () => {
+  let listener;
+  const FakeComponent = () => {
+    const [isUnsaved, setUnsaved] = useState(true);
+
+    useWarnOnPageUnload(isUnsaved);
+
+    return (
+      <button type="button" onClick={() => setUnsaved(false)}>
+        Set not unsaved
+      </button>
+    );
+  };
+  const createComponent = () => mount(<FakeComponent />);
+
+  beforeEach(() => {
+    sinon.stub(window, 'addEventListener').callsFake((eventName, callback) => {
+      listener = callback;
+    });
+    sinon.stub(window, 'removeEventListener');
+  });
+
+  afterEach(() => {
+    window.addEventListener.restore();
+    window.removeEventListener.restore();
+  });
+
+  it('registers event listener when unsaved data is true', () => {
+    createComponent();
+
+    assert.called(window.addEventListener);
+    assert.notCalled(window.removeEventListener);
+  });
+
+  it('unregisters event listener when unsaved data is false', () => {
+    const wrapper = createComponent();
+
+    wrapper.find('button').simulate('click');
+    wrapper.update();
+
+    assert.called(window.removeEventListener);
+  });
+
+  it('unregisters event listener when component is unmounted', () => {
+    const wrapper = createComponent();
+    wrapper.unmount();
+
+    assert.called(window.removeEventListener);
+  });
+
+  it('acts on provided event when listener is invoked', () => {
+    createComponent();
+
+    const event = new Event('beforeupdate');
```

Comment:
The event name here is wrong! (`beforeupdate` instead of `beforeunload`)

To clarify, we're stubbing `beforeunload` because dispatching a real `beforeunload` event causes warnings in the test runner?

A possible solution, rather than stubbing the event handler methods, is to add a test seam for the `window` global. Then in the tests you can use any alternative `EventTarget` instance instead (eg. `fakeWindow = new EventTarget()`).

Reply:
Oh no! 😅 Good catch.

I initially added a test seam for the window, but it felt awkward having to add it as a dependency of `useEffect`, that's why I ended up stubbing the global window object.

But I guess it's not terrible and I like your approach better anyway.